h2/src/main/org/h2/result/LocalResultFactory.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2004-2018 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (http://h2database.com/html/license.html).
+ * Initial Developer: H2 Group
+ */
+package org.h2.result;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import org.h2.engine.Session;
+import org.h2.expression.Expression;
+import org.h2.message.DbException;
+import org.h2.value.DataType;
+import org.h2.value.Value;
+
+/**
+ * Creates local result.
+ */
+public abstract class LocalResultFactory {
+    /**
+     * Default implementation of local result factory.
+     */
+    public static final LocalResultFactory DEFAULT = new DefaultLocalResultFactory();
+
+    /**
+     * Create a local result object.
+     *
+     * @param session the session
+     * @param expressions the expression array
+     * @param visibleColumnCount the number of visible columns
+     * @return object to collect local result.
+     */
+    public abstract LocalResult create(Session session, Expression[] expressions, int visibleColumnCount);
+
+    /**
+     * Create a local result object.
+     * @return object to collect local result.
+     */
+    public abstract LocalResult create();
+
+    /**
+     * Default implementation of local result factory.
+     */
+    static final class DefaultLocalResultFactory extends LocalResultFactory {
+        @Override
+        public LocalResult create(Session session, Expression[] expressions, int visibleColumnCount) {
+            return new LocalResultImpl(session, expressions, visibleColumnCount);
+        }
+
+        @Override
+        public LocalResult create() {
+            return new LocalResultImpl();
+        }
+    }
+
+    /**
+     * Create a local result object by factory registered at the database.
+     *
+     * @param session the session
+     * @param expressions the expression array
+     * @param visibleColumnCount the number of visible columns
+     * @return objetc to collect local result.
+     */
+    public static LocalResult createRow(Session session, Expression[] expressions, int visibleColumnCount) {
+        return session.getDatabase().getLocalResultFactory().create(session, expressions, visibleColumnCount);
```

Comment:
May be simply inline both variants of `createRow()` method into its callers? `db.getLocalResultFactory().create(…)` or `session.getDatabase().getLocalResultFactory().create()` are not significantly large and they will not have confusing and incorrect `createRow()` name.

Also I think that name can be reduced to `Database.getResultFactory()`.

@grandinj, what do you think?

Reply:
Fixed.

But have a small question:
I looked at the `RowFactory` patch like as a pattern. `Database` & `Session` implements `createRow` to simplify row creation at callers. 

Does it make sense to use the same pattern in both cases for style consistency?
I guess the `createRow` row has more performance impact method then `createResult`.

Reply:
Many callers have a `Database` object anyway, I don't see a reason for an additional method in a `Session`.

Reply:
So, what looks better:
- `db.createResult(…)`
- or `db.getLocalResultFactory().create(…)`?

Reply:
I don't see any benefit in additional method, only a few classes use it. I think that you may leave it as is.